tables/schema.py

```diff
@@ -0,0 +1,281 @@
+        return processed_tables
+
+    def validate(self, table: pd.DataFrame) -> tuple[pd.DataFrame | None, list[ErrorMessage] | None]:
+        """Validates a table against the schema using pandera.
+
+        :param table: table to validate
+        :return: validated tables, error messages
+        """
+        validated_table = None
+        error_messages = None
+
+        try:
+            # TODO: currently we won't return tables that don't pass validation, which means any later validation
+            #   i.e. specific_validation wouldn't be executed, leading to a suboptimal "2 stage" validation process
+            #   Can we return the extracted (unvalidated, and therefore un-coerced tables?
+            #   Or should this be left for the calling function to deal with
+            validated_table = self._pandera_schema.validate(table, lazy=True)
+        except pa.errors.SchemaErrors as schema_errors:
+            error_messages = self._handle_errors(schema_errors, table)
+
+        return validated_table, error_messages
+
+    def _handle_errors(self, schema_errors: pa.errors.SchemaErrors, table: pd.DataFrame):
+        """Handle SchemaErrors by converting them to messages that refer to the original spreadsheet.
+
+        :param schema_errors: schema errors found during validation
+        :return: error messages
+        """
+        if not hasattr(table, "header_to_letter"):
+            raise TableExtractError(
+                "Table does not contain header to letter mapping - ensure that the table has not "
+                "been operated on between extract and validate."
+            )
+
+        error_messages = []
+        for failure in schema_errors.failure_cases.itertuples(name="Failure", index=False):
+            error_type = failure.check.split("(")[0]
+
+            # "dtype" check is series-wise and so will not produce element-wise messages. The solution to
+            # this is to use "coerce_dtype" errors instead as these are raised per element.
+            if error_type == "dtype":
+                continue
+
+            if error_type == "column_in_schema":
+                # "column_in_schema" means that the table didn't fit the schema - dev error rather than user
+                # TODO: should we use a different exception?
+                raise TableExtractError("Validated table contains columns from outside of the schema")
+
+            column_letter = table.header_to_letter[failure.column]
+
+            # first try error type + column, then error_type
+            description = self.messages.get((error_type, failure.column)) or self.messages.get(error_type)
+            if not description:
+                raise ValueError(f"No message configured for column {failure.column} with error type {error_type}")
+
+            error_messages.append(
+                ErrorMessage(
+                    self.worksheet_name, self.section, f"{column_letter}{failure.index + 1}", description, error_type
```

Comment:
Why ```failure.index + 1```? Perhaps a comment to explain this as glancing at the code I don't see the reason.

Reply:
This is because Excel files are 1-indexed - will add a comment 👍